### PR TITLE
Remove SCREEN_TEXTURE, DEPTH_TEXTURE, and NORMAL_ROUGHNESS_TEXTURE

### PR DESCRIPTION
**Pull request by clayjohn**
Follow up to https://twitter.com/reduzio/status/1609687033450229760

Reduz and I have discussed this at great length. I was in favour of keeping ``SCREEN_TEXTURE``, ``DEPTH_TEXTURE``, and ``NORMAL_ROUGHNESS_TEXTURE`` as an easy-to-use pathway for screen-space sampling. Reduz preferred to remove them as the new pathway is not complicated and can be much better for performance.

This PR removes ``SCREEN_TEXTURE``, ``DEPTH_TEXTURE``, and ``NORMAL_ROUGHNESS_TEXTURE``, and exposes the new  hint to visual shaders (no changes were needed to text shaders). This PR also cleans up a few bugs related to using these textures.

Now users will always have to use the hint syntax so this:
```
shader_type spatial

void fragment() {
    ALBEDO = texture(SCREEN_TEXTURE, SCREEN_UV).rgb;
}
```
becomes this
```
shader_type spatial

uniform sampler2D screen_texture : hint_screen_texture, repeat_disable, filter_linear_mipmap;

void fragment() {
    ALBEDO = texture(screen_texture, SCREEN_UV).rgb;
}
```
Or for the others:
```
uniform sampler2D depth_texture : hint_depth_texture;
uniform sampler2D normal_roughness_texture : hint_normal_roughness_texture;
```
The repeat mode and filter setting can be whatever you want.

CC @reduz

### CODE REVIEW

**Comment by reduz** [comment]
I think we are good without them. If users complain and make a good case we can add them back, but the new syntax to do this should still be easy enough.

File: scene/resources/visual_shader.cpp [hunk 1]
```diff
@@ -2676,9 +2676,9 @@ const VisualShaderNodeInput::Port VisualShaderNodeInput::ports[] = {
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_VECTOR_2D, "viewport_size", "VIEWPORT_SIZE" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_BOOLEAN, "output_is_srgb", "OUTPUT_IS_SRGB" },
 	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_BOOLEAN, "front_facing", "FRONT_FACING" },
-	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SAMPLER, "screen_texture", "SCREEN_TEXTURE" },
-	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SAMPLER, "normal_roughness_texture", "NORMAL_ROUGHNESS_TEXTURE" },
-	{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SAMPLER, "depth_texture", "DEPTH_TEXTURE" },
+	//{ Shader::MODE_SPATIAL, VisualShader::TYPE_FRAGMENT, VisualShaderNode::PORT_TYPE_SAMPLER, "screen_texture", "SCREEN_TEXTURE" },
```
Comment (reduz) [inline scene/resources/visual_shader.cpp:2679]: Guess these could be removed

**Comment by clayjohn** [comment]
@reduz pushed requested changes and added doc updates for CI

**Review by reduz — APPROVED** [review]

**Comment by clayjohn** [comment]
This PR will conflict with https://github.com/godotengine/godot/pull/71455, so whichever one is merged first the other will need second will need a rebase.

Also, we need to explicitly mention the removal of ``SCREEN_TEXTURE``, ``DEPTH_TEXTURE``, and ``NORMAL_ROUGHNESS_TEXTURE`` in the next Beta release post following this PR

**Comment by BastiaanOlij** [comment]
Indeed, we're going to have to make sure we keep the functionality for making this work in stereo and switching the sampler type.

I have to admit I'm worried about the impact of this especially so close to 4.0s release, but at the same time this also isn't a change we want to introduce halfway through a major release cycle.

I wonder if an intermediate solution would be to at least have a proper warning generated if someone is still using the old syntax, if it just gives a bland syntax error it's going to throw off a lot of people converting their Godot 3 projects or with existing Godot 4 projects if they don't read the release notes carefully.

**Comment by clayjohn** [comment]
@BastiaanOlij Updated to do 2 more things:
1. Remove the dependencies on the old names internally. Now we only check the uniform hint to see its type. This made it easier to avoid errors from things like ``uniform sampler2D SCREEN_TEXTURE : hint_depth_texture;``
2. Add a nice in-editor warning message when users rely on the old names (unfortunately, it gets a little bit cut off, but I think more text is better than less here)

Screenshot from 2023-01-18 09-39-34

**Comment by BastiaanOlij** [comment]
@clayjohn I LOVE IT!

**Comment by akien-mga** [comment]
Looks great!

> unfortunately, it gets a little bit cut off, but I think more text is better than less here

I believe we should enable line wrapping and either let the error panel span more than two lines, or add a vertical scrollbar when needed. But that can be done in a separate PR.

**Comment by akien-mga** [comment]
Thanks!

**Comment by wouldntsavezion** [comment]
So this seems to break everything that used it which is fine except that it also includes the StandardMaterial (eg proximity fade is broken)

**Comment by pixezy** [comment]
SCREEN_TEXTURE just doesn't work. Am I doing something wrong?
uniform sampler2D SCREEN_TEXTURE : hint_screen_texture, repeat_disable, filter_linear_mipmap;

**Comment by owstetra** [comment]
i do keep getting this error :  "Uniforms were never supplied for set (1) at the time of drawing, which are required by the pipeline"
when i replaced : SCREEN_TEXTURE with : screen_texture uniform variable

**Comment by JohnPandich** [comment]
> SCREEN_TEXTURE just doesn't work. Am I doing something wrong? uniform sampler2D SCREEN_TEXTURE : hint_screen_texture, repeat_disable, filter_linear_mipmap;

I did some testing and the issue seems to be with the "hint_screen_texture" hint

**Comment by jonSP12** [comment]
Is there a way to solve this error ?
something to do with the PI ?

redPI

**Comment by akien-mga** [comment]
The error tells you that you are redefining `PI`, which means that `PI` was already defined. It's a built-in constant.

**Comment by jonSP12** [comment]
> The error tells you that you are redefining `PI`, which means that `PI` was already defined. It's a built-in constant.

thanks, changed it to "const float PII = 3.14159265359;"
btw do you know the new script to 'get' and 'set' a shader parameter in gdScript ?

**Comment by akien-mga** [comment]
> thanks, changed it to "const float PII = 3.14159265359;"

My point was that you can simply remove this line... `PI` is already defined so you can simply use it.

For the rest, please use [other community platforms](https://godotengine.org/community/) (e.g. Discord or the Q&A) to ask support questions, this is a merged PR and the discussion should only be about this PR.

**Comment by elvisish** [comment]
> I think we are good without them. If users complain and make a good case we can add them back, but the new syntax to do this should still be easy enough.

I think it makes it awkward to use shaders that were written before GD4, it might even be more convenient if the default shader template contained the hint as so many shaders use it and screen_texture is one of the most important uses of a shader.

**Pull request merged**